## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#  YOUR CODE GOES HERE
application_df = application_df.drop(columns=["EIN"])
application_df.head()


NAME APPLICATION_TYPE  \
0              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Determine the number of unique values in each column.
#  YOUR CODE GOES HERE
application_df.nunique()

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
application__counts = application_df["APPLICATION_TYPE"].value_counts()
application__counts

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64

In [5]:
application__counts[application__counts < 500]

APPLICATION_TYPE
T9     156
T13     66
T12     27
T2      16
T25      3
T14      3
T29      2
T15      2
T17      1
Name: count, dtype: int64

In [6]:
application__counts[application__counts < 500].index

Index(['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17'], dtype='object', name='APPLICATION_TYPE')

In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
application_types_to_replace = list(application__counts[application__counts < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
#  YOUR CODE GOES HERE
class_counts = application_df["CLASSIFICATION"].value_counts()
class_counts

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [9]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#  YOUR CODE GOES HERE
class_counts[class_counts > 1]
class_counts

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [10]:
class_counts[class_counts > 10]
class_counts

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [11]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
#  YOUR CODE GOES HERE
classifications_to_replace = list(class_counts[class_counts < 1000].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
#  YOUR CODE GOES HERE
application_with_dummies_df = pd.get_dummies(application_df)
application_with_dummies_df

STATUS   ASK_AMT  IS_SUCCESSFUL  \
0           1      5000              1   
1           1    108590              1   
2           1      5000              0   
3           1      6692              1   
4           1    142590              1   
...       ...       ...            ...   
34294       1      5000              0   
34295       1      5000              0   
34296       1      5000              0   
34297       1      5000              1   
34298       1  36500179              0   

       NAME_1 DAY RANCH RESCUE AND RURAL OKLAHOMA ANIMAL RESOURCE INC  \
0                                                  False                
1                                                  False                
2                                                  False                
3                                                  False                
4                                                  False                
...                                                  ...                
34294                                              False                
34295                                              False                
34296                                              False                
34297                                              False                
34298                                              False                

       NAME_100 BLACK MEN OF AMERICA  NAME_100 BLACK MEN OF MEMPHIS INC  \
0                              False                              False   
1                              False                              False   
2                              False                              False   
3                              False                              False   
4                              False                              False   
...                              ...                                ...   
34294                          False                              False   
34295                          False                              False   
34296                          False                              False   
34297                          False                              False   
34298                          False                              False   

       NAME_100 BLACK MEN OF WEST GEORGIA INC  NAME_1150 WEBSTER STREET INC  \
0                                       False                         False   
1                                       False                         False   
2                                       False                         False   
3                                       False                         False   
4                                       False                         False   
...                                       ...                           ...   
34294                                   False                         False   
34295                                   False                         False   
34296                                   False                         False   
34297                                   False                         False   
34298                                   False                         False   

       NAME_116TH CAVALRY REGIMENT CHAPTER OF THE US CAVALRY & ARMOR ASSOCIATION  \
0                                                  False                           
1                                                  False                           
2                                                  False                           
3                                                  False                           
4                                                  False                           
...                                                  ...                           
34294                                              False                           
34295                                              False                           
34296              

In [13]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE
X = application_with_dummies_df.drop(["IS_SUCCESSFUL"], axis="columns").values
y = application_with_dummies_df["IS_SUCCESSFUL"].values

# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE
number_input_feature = len(X_train[0])
hidden_nodes_layers1 = 100
hidden_nodes_layers2 = 30
hidden_nodes_layers3 = 10


nn = tf.keras.models.Sequential()

# First hidden layer
#  YOUR CODE GOES HERE
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layers1, input_dim=number_input_feature, activation="relu")
)

# Second hidden layer
#  YOUR CODE GOES HERE
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layers2, activation="sigmoid")
)

nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layers3, activation="sigmoid")
)




# Output layer
#  YOUR CODE GOES HERE
nn.add(
    tf.keras.layers.Dense(units=1, activation="sigmoid")
)

# Check the structure of the model
nn.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               1961200   
                                                                 
 dense_1 (Dense)             (None, 30)                3030      
                                                                 
 dense_2 (Dense)             (None, 10)                310       
                                                                 
 dense_3 (Dense)             (None, 1)                 11        
                                                                 
Total params: 1964551 (7.49 MB)
Trainable params: 1964551 (7.49 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [16]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [17]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100


804/804 [==============================] - 11s 11ms/step - loss: 0.5803 - accuracy: 0.7366
Epoch 2/100
804/804 [==============================] - 9s 11ms/step - loss: 0.3854 - accuracy: 0.8366
Epoch 3/100
804/804 [==============================] - 8s 10ms/step - loss: 0.1784 - accuracy: 0.9414
Epoch 4/100
804/804 [==============================] - 8s 10ms/step - loss: 0.1118 - accuracy: 0.9621
Epoch 5/100
804/804 [==============================] - 8s 10ms/step - loss: 0.1038 - accuracy: 0.9627
Epoch 6/100
804/804 [==============================] - 8s 10ms/step - loss: 0.0978 - accuracy: 0.9628
Epoch 7/100
804/804 [==============================] - 8s 10ms/step - loss: 0.0953 - accuracy: 0.9621
Epoch 8/100
804/804 [==============================] - 8s 10ms/step - loss: 0.0928 - accuracy: 0.9628
Epoch 9/100
804/804 [==============================] - 9s 11ms/step - loss: 0.0915 - accuracy: 0.9635
Epoch 10/100
804/804 [==============================] - 8s 10ms/step - loss: 0.

804/804 [==============================] - 8s 10ms/step - loss: 0.0952 - accuracy: 0.9625
Epoch 77/100
804/804 [==============================] - 8s 10ms/step - loss: 0.0934 - accuracy: 0.9621
Epoch 78/100
804/804 [==============================] - 8s 10ms/step - loss: 0.0927 - accuracy: 0.9623
Epoch 79/100
804/804 [==============================] - 10s 12ms/step - loss: 0.0913 - accuracy: 0.9620
Epoch 80/100
804/804 [==============================] - 9s 12ms/step - loss: 0.0916 - accuracy: 0.9628
Epoch 81/100
804/804 [==============================] - 9s 11ms/step - loss: 0.0922 - accuracy: 0.9628
Epoch 82/100
804/804 [==============================] - 9s 11ms/step - loss: 0.0907 - accuracy: 0.9630
Epoch 83/100
804/804 [==============================] - 9s 11ms/step - loss: 0.0903 - accuracy: 0.9641
Epoch 84/100
804/804 [==============================] - 9s 11ms/step - loss: 0.0916 - accuracy: 0.9635
Epoch 85/100
804/804 [==============================] - 9s 11ms/step - loss: 0.0915 -

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5678 - accuracy: 0.7352 - 787ms/epoch - 3ms/step
Loss: 0.5678457617759705, Accuracy: 0.7351603507995605


In [19]:
# Export our model to HDF5 file
#  YOUR CODE GOES HERE
nn.save("AlphabetSoupCharity.h5")

C:\Users\kento\anaconda3\Lib\site-packages\keras\src\engine\training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


# The report should contain the following:

# Overview of the analysis: Explain the purpose of this analysis.

# Results: Using bulleted lists and images to support your answers, address the following questions:

# Data Preprocessing

1. What variable(s) are the target(s) for your model?
   
   Answer: The target is the Is-Successful column

2. What variable(s) are the features for your model?
   
   Answer: These are:
                    APPLICATION_TYPE
                    AFFILIATION
                    CLASSIFICATION
                    USE_CASE
                    ORGANIZATION
                    STATUS
                    INCOME_AMT
                    SPECIAL_CONSIDERATIONS
                    ASK_AMT

3. What variable(s) should be removed from the input data because they are neither targets nor features?

   Answer: EIN

# Compiling, Training, and Evaluating the Model

1. How many neurons, layers, and activation functions did you select for your neural network model, and why?

  Answer: I used 3 hidden layers hoping to increase the accuracy from "a" solution at 72%.  However, it did not turn out to be a small improvement to 73.5%.  

2. Were you able to achieve the target model performance?

  Anser: Not as expected to be above 75%.

3. What steps did you take in your attempts to increase model performance?

  Answer: Kept NAME data points since it has has the biggest counts and hoped to inprove the accuracy.

# Summary: Summarize the overall results of the deep learning model.

I could not achieve the improvement in this model as expected.  Colab kept giving me RAM issue many times and I was unable to run this model in Colab.  I was advised to run this model locally. It took forever to train the model.  

In [20]:
from sklearn.metrics import accuracy_scorefrom
from sklearn.ensemble import RandomForestClassificatier

ImportError: cannot import name 'accuracy_scorefrom' from 'sklearn.metrics' (C:\Users\kento\anaconda3\Lib\site-packages\sklearn\metrics\__init__.py)

In [21]:
rf_model = RandomForestClassificatier(n_estimators=120, random_state=78)

rf_model = rf+_model.fit(X_train_scaled, y_train)

y_pred = rf_model.predict(X_test_scaled)

print(f"Random forest model accuracy {accuracy_score(y_test, y_pred)}")



NameError: name 'RandomForestClassificatier' is not defined